larpandoracontent/LArTwoDReco/LArCosmicRay/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,877 @@
+        return false;
+
+    const CartesianVector predictedUpstreamPoint(downstreamPoint + (downstreamDirection * separationDistance));
+    const float predictedUpstreamOffset((predictedUpstreamPoint - upstreamPoint).GetMagnitude());
+
+    if (predictedUpstreamOffset > m_maxPredictedMergePointOffset)
+        return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::GetExtrapolatedCaloHits(const ClusterAssociation &clusterAssociation, const ClusterList *const pClusterList,
+    ClusterToCaloHitListMap &clusterToCaloHitListMap) const
+{
+    const CartesianVector &upstreamPoint(clusterAssociation.GetUpstreamMergePoint()), &downstreamPoint(clusterAssociation.GetDownstreamMergePoint());
+    const float minX(std::min(upstreamPoint.GetX(), downstreamPoint.GetX())), maxX(std::max(upstreamPoint.GetX(), downstreamPoint.GetX()));
+
+    for (const Cluster *const pCluster : *pClusterList)
+    {
+        const OrderedCaloHitList &orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+        for (const OrderedCaloHitList::value_type &mapEntry : orderedCaloHitList)
+        {
+            for (const CaloHit *const pCaloHit : *mapEntry.second)
+            {
+                const CartesianVector &hitPosition(pCaloHit->GetPositionVector());
+                if ((hitPosition.GetX() < minX) || (hitPosition.GetX() > maxX) || (hitPosition.GetZ() < upstreamPoint.GetZ()) || (hitPosition.GetZ() > downstreamPoint.GetZ()))
+                    continue;
+
+                const float distanceFromLine(clusterAssociation.GetConnectingLineDirection().GetCrossProduct(hitPosition - upstreamPoint).GetMagnitude());
+                if (distanceFromLine > m_distanceFromLine)
+                    continue;
+
+                clusterToCaloHitListMap[pCluster].push_back(pCaloHit);
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TrackInEMShowerAlgorithm::IsTrackContinuous(const ClusterAssociation &clusterAssociation, const ClusterToCaloHitListMap &clusterToCaloHitListMap) const
+{
+    // ATTN: Collect extrapolated calo hits and sort by projected distance from the upstreamMergePoint
+    CaloHitVector extrapolatedCaloHitVector;
+    for (const auto &entry : clusterToCaloHitListMap)
+        extrapolatedCaloHitVector.insert(extrapolatedCaloHitVector.begin(), entry.second.begin(), entry.second.end());
+
+    std::sort(extrapolatedCaloHitVector.begin(), extrapolatedCaloHitVector.end(), SortByDistanceAlongLine(clusterAssociation.GetUpstreamMergePoint(),
+        clusterAssociation.GetConnectingLineDirection()));
+
+    CartesianPointVector trackSegmentBoundaries;
+    this->GetTrackSegmentBoundaries(clusterAssociation, trackSegmentBoundaries);
+
```

Comment:
Worth having a check on `trackSegmentBoundaries.size()`, as below subtract 1 from size, and use `.at(i)` and `.at(i + 1)`. Can just return true/false as appropriate?